File: rust/src/policy/export.rs
//! Policy training data export (A4.4).
//!
//! Exports decision-outcome pairs as JSONL files for offline policy model training.
//! Requires sufficient assessed data before export is meaningful.

use std::io::Write;
use std::path::PathBuf;

use log::info;

use crate::config::{default_policy_data_dir, MemoryConfig};
use crate::storage::sqlite_store::SQLiteStore;

/// Result of a policy data export operation.
#[derive(Debug, Clone)]
pub struct ExportResult {
    pub save_examples: usize,
    pub retrieval_examples: usize,
    pub save_path: String,
    pub retrieval_path: String,
    pub ready_for_training: bool,
}

/// Export decision-outcome pairs for offline policy model training.
///
/// Returns metadata about the exported data: example counts and file paths.
pub async fn export_policy_training_data(
    sqlite: &SQLiteStore,
    config: &MemoryConfig,
    output_dir: Option<PathBuf>,
) -> Result<ExportResult, Box<dyn std::error::Error + Send + Sync>> {
    let output_dir = output_dir.unwrap_or_else(default_policy_data_dir);
    std::fs::create_dir_all(&output_dir)?;

    let save_data = sqlite.export_save_policy_data().await?;
    let retrieval_data = sqlite.export_retrieval_policy_data().await?;

    let save_path = output_dir.join("save_policy_data.jsonl");
    let retrieval_path = output_dir.join("retrieval_policy_data.jsonl");

    {
        let mut f = std::fs::File::create(&save_path)?;
        for row in &save_data {
            let line = serde_json::to_string(row)?;
            writeln!(f, "{line}")?;
        }
    }

    {
        let mut f = std::fs::File::create(&retrieval_path)?;
        for row in &retrieval_data {
            let line = serde_json::to_string(row)?;
            writeln!(f, "{line}")?;
        }
    }

    let result = ExportResult {
        save_examples: save_data.len(),
        retrieval_examples: retrieval_data.len(),
        save_path: save_path.to_string_lossy().into_owned(),
        retrieval_path: retrieval_path.to_string_lossy().into_owned(),
        ready_for_training: save_data.len() >= config.policy_min_save_examples
            && retrieval_data.len() >= config.policy_min_retrieval_examples,
    };

    info!(
        "Policy data export: {} save examples, {} retrieval examples (ready={})",
        result.save_examples, result.retrieval_examples, result.ready_for_training
    );

    Ok(result)
}

File: go/policy/export.go
package policy

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"

	agentmemory "github.com/CodeHalwell/Memories/go"
	"github.com/CodeHalwell/Memories/go/storage"
)

// ExportResult holds metadata about a policy training data export.
type ExportResult struct {
	SaveExamples      int    `json:"save_examples"`
	RetrievalExamples int    `json:"retrieval_examples"`
	SavePath          string `json:"save_path"`
	RetrievalPath     string `json:"retrieval_path"`
	ReadyForTraining  bool   `json:"ready_for_training"`
}

// ExportPolicyTrainingData exports decision-outcome pairs for offline policy model training (A4.4).
func ExportPolicyTrainingData(ctx context.Context, sqlite *storage.SQLiteStore, cfg agentmemory.MemoryConfig, outputDir *string) (ExportResult, error) {
	dir := cfg.PolicyDataDir
	if outputDir != nil {
		dir = *outputDir
	}
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return ExportResult{}, fmt.Errorf("creating output dir: %w", err)
	}

	saveData, err := sqlite.ExportSavePolicyData(ctx)
	if err != nil {
		return ExportResult{}, fmt.Errorf("exporting save data: %w", err)
	}

	retrievalData, err := sqlite.ExportRetrievalPolicyData(ctx)
	if err != nil {
		return ExportResult{}, fmt.Errorf("exporting retrieval data: %w", err)
	}

	savePath := filepath.Join(dir, "save_policy_data.jsonl")
	retrievalPath := filepath.Join(dir, "retrieval_policy_data.jsonl")

	if err := writeJSONL(savePath, saveData); err != nil {
		return ExportResult{}, err
	}
	if err := writeJSONL(retrievalPath, retrievalData); err != nil {
		return ExportResult{}, err
	}

	result := ExportResult{
		SaveExamples:      len(saveData),
		RetrievalExamples: len(retrievalData),
		SavePath:          savePath,
		RetrievalPath:     retrievalPath,
		ReadyForTraining: len(saveData) >= cfg.PolicyMinSaveExamples &&
			len(retrievalData) >= cfg.PolicyMinRetrievalExamples,
	}

	log.Printf("Policy data export: %d save examples, %d retrieval examples (ready=%v)",
		result.SaveExamples, result.RetrievalExamples, result.ReadyForTraining)

	return result, nil
}

func writeJSONL(path string, rows []map[string]interface{}) error {
	f, err := os.Create(path)
	if err != nil {
		return fmt.Errorf("creating %s: %w", path, err)
	}
	defer f.Close()

	encoder := json.NewEncoder(f)
	encoder.SetEscapeHTML(false)
	for _, row := range rows {
		if err := encoder.Encode(row); err != nil {
			return fmt.Errorf("writing to %s: %w", path, err)
		}
	}
	return nil
}
